src/main/scala/Util.scala

```diff
@@ -59,11 +61,19 @@ case class AsyncWideCounter(width: Int, inc: UInt = 1.U, reset: Boolean = true)
 
 // As WideCounter, but it's a module so it can take arbitrary clocks
 class WideCounterModule(w: Int, inc: UInt = 1.U, reset: Boolean = true, clockSignal: Clock = null, resetSignal: Bool = null)
-    extends Module(Option(clockSignal), Option(resetSignal)) {
+    extends Module {
   val io = IO(new Bundle {
     val value = Output(UInt(w.W))
   })
-  io.value := AsyncWideCounter(w, inc, reset).value
+  lazy val block = {
+    io.value := AsyncWideCounter(w, inc, reset).value
+  }
+  (clockSignal, resetSignal) match {
+    case (null, null) => block
+    case (null, r: Bool) => withReset(r) { block }
+    case (c: Clock, null) => withClock(c) { block }
+    case (c: Clock, r: Bool) => withClockAndReset(c, r) { block }
+  }
```

Comment:
Much better. But there's potential confusion with `reset` since it's a parameter to this class. How about:
```scala
  io.value := {
    val cc = Option(clockSignal).getOrElse(super.clock)
    val rr = Option(resetSignal).getOrElse(super.reset)
    withClockAndReset(cc, rr) { AsyncWideCounter(w, inc, reset).value }
  }
```

Reply:
Nope.
```
Error:(69, 50) super may not be used on value clock
    val cc = Option(clockSignal).getOrElse(super.clock)
Error:(70, 50) super may not be used on value reset
    val rr = Option(resetSignal).getOrElse(super.reset)
```

Reply:
The _obvious_ solution:
```scala
class WideCounterModule(w: Int, inc: UInt = 1.U, doReset: Boolean = true, clockSignal: Clock = null, resetSignal: Bool = null)
    extends Module {
  val io = IO(new Bundle {
    val value = Output(UInt(w.W))
  })
  io.value := {
    val cc = Option(clockSignal).getOrElse(this.clock)
    val rr = Option(resetSignal).getOrElse(this.reset)
    withClockAndReset(cc, rr) { AsyncWideCounter(w, inc, doReset).value }
  }
}
```